Unit Test

rpmartz opened the pull request:
Now that we have Maven, we can start using third party libraries. One library that nearly every project uses is one called [JUnit](http://junit.org/). It is a testing library that lets you add automated tests that Maven runs every time it builds. 

We'll talk a lot more about unit testing as we go on with the project, but at it's heart it's about making sure the code you right is doing what it's supposed to do. For example, we're going to write tests that our `TransferManager`'s `transfer` method transfers the correct amount of money to the correct account.  

This will seem a little bit silly at first, because you're like well we can just run the `main` method to and verify that it worked, but that doesn't work when you have hundreds or thousands of different classes that consist of hundreds of thousands of lines of code. 

## Running Tests

If you go to the project directory (the same directory that has the `pom.xml`) and run `mvnw.cmd test` or `mvnw.cmd package` you will see that the build fails. One of the tests pass and one of them doesn't (your homework is to make the one that is failing pass). 

Check out the files changed for specifics.

rpmartz committed:
move files to src/main/java structure maven expects

rpmartz committed:
add target/ to .gitignore

rpmartz committed:
add maven wrapper

rpmartz committed:
add basic POM.xml

rpmartz committed:
set source and target properties of the compiler plugin to 1.8

rpmartz committed:
add junit

rpmartz committed:
add example passing test

rpmartz commented on src/test/java/com/patrickjones/services/TransferManagerTest.java:
```diff
@@ -0,0 +1,32 @@
+
+import com.patrickjones.models.BancAccount;
+
+public class TransferManagerTest {
```
This is a unit test class. Notice that it is the name of the class being tested (`TransferManager`) + `Test`. These files have to either begin or end with the word `Test` so that Maven knows to run them when it runs the unit tests.

rpmartz commented on src/test/java/com/patrickjones/services/TransferManagerTest.java:
```diff
@@ -0,0 +1,32 @@
+
+public class TransferManagerTest {
+
+	@Test
```
`@Test` is an annotation that lets JUnit know to run that method as a unit test.

rpmartz commented on src/test/java/com/patrickjones/services/TransferManagerTest.java:
```diff
@@ -0,0 +1,32 @@
+		transferManager.transfer(acctToDebit, acctToCredit, 100);
+
+		String failureMessage = "The account to debit's balance should be reduced by the amount to transfer";
+		assertEquals(failureMessage, 900, acctToDebit.getBalance());
```
Most of the time you will use JUnit's `Assert` classes to test your code. In this case, we're testing that the `accountToDebit`'s balance gets reduced by `100` since that is what we're transferring. 

Notice the format of the test -we set up some dummy `BancAccount` classes, set up the class we want to test (the `TransferManager` in this case), then call the method that does the action we want to test (`transfer`) in this case. After we call `transfer` we perform the assertions to test what we want to ensure either happened or didn't happen.

rpmartz commented on src/test/java/com/patrickjones/services/TransferManagerTest.java:
```diff
@@ -0,0 +1,32 @@
+
+	@Test
+	public void testCreditsAccountToCreditByTransferAmount() {
+		Assert.fail("Patrick has to complete me");
```
You need to replace this line with a test similar to the test in the `testDebitsAccountToDebitByTransferAmount` above this, but this time you need to test that the `accountToCredit` gets its balance reduced by however much you transfer.

rpmartz commented:
@patrickatcw I should have also mentioned that if you open up the `TransferManagerTest` in IntelliJ you can run the tests from there as well.

patrickatcw commented:
Hey Ryan,

Fixed testing error and updated branch to master.

Patrick
--------------------------------------------
On Tue, 2/13/18, Ryan Martz <notifications@github.com> wrote:

 Subject: Re: [patrickatcw/JavaBanco] Unit Test (#6)
 To: "patrickatcw/JavaBanco" <JavaBanco@noreply.github.com>
 Cc: "patrickatcw" <pj8392@yahoo.com>, "Mention" <mention@noreply.github.com>
 Date: Tuesday, February 13, 2018, 9:03 PM
 
 @patrickatcw I should
 have also mentioned that if you open up the
 TransferManagerTest in IntelliJ you can run the
 tests from there as well.
 
 —
 You are receiving this because you were mentioned.
 Reply to this email directly, view
 it on GitHub, or mute
 the thread.

rpmartz commented:
Beautiful, looking good man. I added two more tasks for you. Let me know if you get stuck on anything.

> On Feb 19, 2018, at 3:25 PM, patrickatcw <notifications@github.com> wrote:
> 
> Hey Ryan,
> 
> Fixed testing error and updated branch to master.
> 
> Patrick
> --------------------------------------------
> On Tue, 2/13/18, Ryan Martz <notifications@github.com> wrote:
> 
> Subject: Re: [patrickatcw/JavaBanco] Unit Test (#6)
> To: "patrickatcw/JavaBanco" <JavaBanco@noreply.github.com>
> Cc: "patrickatcw" <pj8392@yahoo.com>, "Mention" <mention@noreply.github.com>
> Date: Tuesday, February 13, 2018, 9:03 PM
> 
> @patrickatcw I should
> have also mentioned that if you open up the
> TransferManagerTest in IntelliJ you can run the
> tests from there as well.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view
> it on GitHub, or mute
> the thread.
> 
> 
> 
> 
> 
> 
> 
> 
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub <https://github.com/patrickatcw/JavaBanco/pull/6#issuecomment-366794953>, or mute the thread <https://github.com/notifications/unsubscribe-auth/ABq2cZMnXaYJDpGf4tsiBmGlJvxGc-RTks5tWdisgaJpZM4SBhqR>.
>

patrickatcw commented:
ok, thanks buddy!  will work on it today, along with some other things i am working on.  catch up later!

Patrick
--------------------------------------------
On Tue, 2/20/18, Ryan Martz <notifications@github.com> wrote:

 Subject: Re: [patrickatcw/JavaBanco] Unit Test (#6)
 To: "patrickatcw/JavaBanco" <JavaBanco@noreply.github.com>
 Cc: "patrickatcw" <pj8392@yahoo.com>, "Mention" <mention@noreply.github.com>
 Date: Tuesday, February 20, 2018, 10:34 AM
 
 Beautiful, looking good man. I added
 two more tasks for you. Let me know if you get stuck on
 anything.
 
 
 
 > On Feb 19, 2018, at 3:25 PM, patrickatcw
 <notifications@github.com> wrote:
 
 > 
 
 > Hey Ryan,
 
 > 
 
 > Fixed testing error and updated branch to master.
 
 > 
 
 > Patrick
 
 > --------------------------------------------
 
 > On Tue, 2/13/18, Ryan Martz
 <notifications@github.com> wrote:
 
 > 
 
 > Subject: Re: [patrickatcw/JavaBanco] Unit Test (#6)
 
 > To: "patrickatcw/JavaBanco"
 <JavaBanco@noreply.github.com>
 
 > Cc: "patrickatcw" <pj8392@yahoo.com>,
 "Mention" <mention@noreply.github.com>
 
 > Date: Tuesday, February 13, 2018, 9:03 PM
 
 > 
 
 > @patrickatcw I should
 
 > have also mentioned that if you open up the
 
 > TransferManagerTest in IntelliJ you can run the
 
 > tests from there as well.
 
 > 
 
 > —
 
 > You are receiving this because you were mentioned.
 
 > Reply to this email directly, view
 
 > it on GitHub, or mute
 
 > the thread.
 
 > 
 
 > 
 
 > 
 
 > 
 
 > 
 
 > 
 
 > 
 
 > 
 
 > 
 
 > —
 
 > You are receiving this because you authored the
 thread.
 
 > Reply to this email directly, view it on GitHub
 <https://github.com/patrickatcw/JavaBanco/pull/6#issuecomment-366794953>,
 or mute the thread
 <https://github.com/notifications/unsubscribe-auth/ABq2cZMnXaYJDpGf4tsiBmGlJvxGc-RTks5tWdisgaJpZM4SBhqR>.
 
 > 
 
 
 
 
 
 —
 You are receiving this because you were mentioned.
 Reply to this email directly, view
 it on GitHub, or mute
 the thread.
